src/csim/update_ops_control_multi_target_multi.cpp

```diff
@@ -67,3 +72,127 @@ void multi_qubit_control_multi_qubit_dense_matrix_gate(
     free(buffer);
     free(matrix_mask_list);
 }
+
+#ifdef _USE_MPI
+void multi_qubit_control_multi_qubit_dense_matrix_gate_mpi(
+    const UINT* control_qubit_index_list, const UINT* control_value_list,
+    UINT control_count, const UINT* target_qubit_index_list, UINT target_count,
+    const CTYPE* matrix, CTYPE* state, ITYPE dim, UINT inner_qc) {
+    UINT* index_list_buf = (UINT*)malloc(
+        (size_t)(sizeof(UINT) * (control_count * 2 + target_count)));
+
+    UINT* new_control_index_list = index_list_buf;
+    UINT* new_control_value_list = index_list_buf + control_count;
+    UINT* new_target_index_list = index_list_buf + control_count * 2;
+
+    // copy control qubit lists
+    for (UINT j = 0; j < control_count; ++j) {
+        new_control_index_list[j] = control_qubit_index_list[j];
+        new_control_value_list[j] = control_value_list[j];
+    }
+    // count outer-target qubits
+    UINT num_outer_target = 0;
+    for (UINT i = 0; i < target_count; ++i) {
+        new_target_index_list[i] = target_qubit_index_list[i];
+        if (target_qubit_index_list[i] >= inner_qc) ++num_outer_target;
+    }
+
```

Comment:
Throw exception if num_outer_target > inner_qc.

Reply:
Thank you. I added an exception here as well.